### PR TITLE
Initial broadcom port for Raspberry Pi

### LINKED ISSUE #4314: BCM2711 port (for Raspberry Pi 4 support)
We should port CircuitPython to the Cortex-A class Raspberry Pis. The main early benefit would be easy HDMI capable displayio.

Prerequisites:

- [x] Port TinyUSB to the BCM chips. (Should be relatively easy because it uses Synopsis USB IP.)
- [x] Add internal flash support for the emmc peripheral of the BCM

This should get CP booting and the workflow working. Then it'll be the common steps or implementing standard hardware APIs like I2C and `displayio`.

### PR DESCRIPTION
**Pull request by tannewt**
This targets the 64-bit CPU Raspberry Pis. The BCM2711 on the Pi 4
and the BCM2837 on the Pi 3 and Zero 2W. There are 64-bit fixes
outside of the ports directory for it.

There are a couple other cleanups that were incidental:
* Use const mcu_pin_obj_t instead of omitting the const. The structs
  themselves are const because they are in ROM.
* Use PTR <-> OBJ conversions in more places. They were found when
  mp_obj_t was set to an integer type rather than pointer.
* Optimize submodule checkout because the Pi submodules are heavy
  and unnecessary for the vast majority of builds.

Fixes #4314

### CODE REVIEW

**Comment by tannewt** [comment]
These are temporary USB PIDs under the Raspberry Pi VID. I'm waiting to hear from @ghollingworth about the official values.

File: shared-module/displayio/Bitmap.h [hunk 1]
```diff
@@ -37,8 +37,8 @@ typedef struct {
     mp_obj_base_t base;
     uint16_t width;
     uint16_t height;
-    size_t *data;
```
Comment (jepler) [inline shared-module/displayio/Bitmap.h:40]: I agree with this change

**Review by dhalbert — CHANGES_REQUESTED** [review]
Amazing work! Lots of good cleanup removing 32-bit assumptions.

File: ports/broadcom/boards/raspberrypi_zero2w/pins.c [hunk 2]
```diff
@@ -0,0 +1,11 @@
+#include "shared-bindings/board/__init__.h"
+
+STATIC const mp_rom_map_elem_t board_global_dict_table[] = {
+    CIRCUITPYTHON_BOARD_DICT_STANDARD_ITEMS
+
+    { MP_ROM_QSTR(MP_QSTR_GPIO18), MP_ROM_PTR(&pin_GPIO18) },
+    { MP_ROM_QSTR(MP_QSTR_GPIO19), MP_ROM_PTR(&pin_GPIO19) },
+    { MP_ROM_QSTR(MP_QSTR_GPIO20), MP_ROM_PTR(&pin_GPIO20) },
+    { MP_ROM_QSTR(MP_QSTR_GPIO21), MP_ROM_PTR(&pin_GPIO21) },
```
[the remark this answers is not in the document]
Comment (tannewt) [reply ports/broadcom/boards/raspberrypi_zero2w/pins.c:9]: I just hadn't copied them over. All of the pinouts for the official boards are essentially the same.

File: py/enum.h [hunk 3]
```diff
@@ -55,11 +55,11 @@ typedef struct {
         { &mp_type_type }, \
         .name = MP_QSTR_##type, \
         .print = typename##_print, \
-        .locals_dict = (mp_obj_t)&typename##_locals_dict, \
+        .locals_dict = (mp_obj_dict_t *)&typename##_locals_dict, \
```
[the remark this answers is not in the document]
Comment (tannewt) [reply py/enum.h:58]: It only warns when mp_obj_t isn't a pointer type. (I had tried repr D before realizing it was weird 64 bits on 32 bit platform.)

File: ports/broadcom/common-hal/microcontroller/__init__.h [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#ifndef MICROPY_INCLUDED_BROADCOM_COMMON_HAL_MICROCONTROLLER___INIT___H
+#define MICROPY_INCLUDED_BROADCOM_COMMON_HAL_MICROCONTROLLER___INIT___H
+
+#define TOTAL_GPIO_COUNT 10
```
[the remark this answers is not in the document]
Comment (tannewt) [reply ports/broadcom/common-hal/microcontroller/__init__.h:30]: Look like it's for the microcontroller.pin array.

File: ports/broadcom/common-hal/busio/UART.c [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+
+#define UART_INST(uart) (((uart) ? uart1 : uart0))
+
+typedef enum {
+    STATUS_FREE = 0,
+    STATUS_BUSY,
+    STATUS_NEVER_RESET
+} uart_status_t;
+
+#define NUM_UARTS 2
```
[the remark this answers is not in the document]
Comment (tannewt) [reply ports/broadcom/common-hal/busio/UART.c:52]: Added a comment. All of the other uarts are a different peripheral type that isn't handled yet.

**Comment by tannewt** [comment]
Ok, I've pushed a new version. It includes I2C support too.

**Comment by dhalbert** [comment]
Thanks for the updates, and hurrah for I2C! zero2w port is failing for some compilation reason, but the pi4b is working.

**Comment by dhalbert** [comment]
> Thanks for the updates, and hurrah for I2C! zero2w port is failing for some compilation reason, but the pi4b is working.

Looks like a simple fix: I added a `#include`. If this works I will merge.

**Comment by dhalbert** [comment]
@tannewt Making a commit from the GitHub web interface seems to have upset the build cloning optimizations you added. From https://github.com/adafruit/circuitpython/runs/4313459204?check_suite_focus=true:

```
Run git describe --dirty --tags
  git describe --dirty --tags
  echo >>$GITHUB_ENV CP_VERSION=$(git describe --dirty --tags)
  shell: /usr/bin/bash -e {0}
  env:
    pythonLocation: /opt/hostedtoolcache/Python/3.8.12/x64
fatal: No tags can describe 'e98223cbae0c9dfe636badcc49bb85fa3f905cfa'.
Try --always, or create some tags.
Error: Process completed with exit code 128.
```

**Comment by dhalbert** [comment]
@tannewt I pushed an empty commit from your repo, and it didn't fix the problem. Hmm.

**Review by dhalbert — APPROVED** [review]

**Pull request merged**